### ISSUE TITLE
CUDA_ERROR_ILLEGAL_ADDRESS (700) when running gentle equilibration on a large system with many contexts

### ISSUE THREAD

**Issue by zhang-ivy** [openmm/openmm#3414]
I tried implementing the gentle equilibration protocol below in OpenMM:
<img width="925" alt="Screen Shot 2022-01-14 at 4 33 12 PM" src="https://user-images.githubusercontent.com/35546250/149588135-8c044513-cccb-45a4-8f4c-ef9485256081.png">

The code below is running fine on alanine dipeptide in solvent, but I get `OpenMMException: Error downloading array buffer: CUDA_ERROR_ILLEGAL_ADDRESS (700)` during stage 4 when running on RBD:ACE2 (200 K atoms).

If I skip stages 2 and 3 (aka run stage 1 and then 4), it runs fine.

If I run all stages (without a restraint force in any of the stages), I get `Particle coordinate is nan`.
And if I run stage 1 and then 4 (without a restraint force in any of the stages), it runs fine.

Any ideas why I'm seeing the CUDA error? Since I am not seeing the error when I skip stages 2 and 3 (or when I use a smaller system), I wonder if this problem arises because I am not deleting contexts appropriately? Or maybe something about stages 2 and 3 is causing an instability?

Here is my code:
```python

def run_gentle_equilibration(topology, positions, system, stages, filename, platform_name='CUDA'):
    """
    Run gentle equilibration.
    
    Parameters
    ----------
    topology : openmm.app.Topology
        topology
    positions : np.array in unit.nanometer
        positions
    system : openmm.System
        system
    stages : list of dicts
        each list corresponds to a stage of equiilbration
        each dict contains the equilibration parameters for a given stage
        
        equilibration parameters:
            EOM : str
                'minimize' or 'MD' or 'MD_interpolate' (the last one will allow interpolation between 'temperature' and 'temperature_end')
            n_steps : int
                number of steps of MD 
            temperature : int
                in units of kelvin
            temperature_end : int, optional
                the temperature at which to finish interpolation, if 'EOM' is 'MD_interpolate'
            ensemble : str or None
                'NPT' or 'NVT'
            restraint_selection : str or None
                to be used by mdtraj to select atoms for which to apply restraints
            force_constant : int
                in units of kcal/molA^2
            collision_rate : int
                 in units of 1/picoseconds
    filename : str
        path to save the equilibrated structure
        
    """
            
    for i, parameters in enumerate(stages):
        
        initial_time = time.time()
        print(f"Executing stage {i + 1}")
        
        # Make a copy of the system
        system_copy = copy.deepcopy(system)
        
        # Set barostat update interval to 0 (for NVT)
        if parameters['ensemble'] == 'NVT':
            force_dict = {force.__class__.__name__: index for index, force in enumerate(system_copy.getForces())}
#             force_dict['MonteCarloBarostat'].setFrequency(0)
            system_copy.removeForce(force_dict['MonteCarloBarostat'])
    
        elif parameters['ensemble'] == 'NPT' or parameters['ensemble'] is None:
            pass
        
        else:
            raise Exception("Invalid parameter supplied for 'ensemble'")
            
        # Set up integrator
        temperature = parameters['temperature'] # kelvin
        collision_rate = parameters['collision_rate'] # 1 / unit.picoseconds
        timestep = 0.002 if i == len(stages) - 1 else 0.001 # picoseconds
        
        if parameters['EOM'] == 'MD_interpolate':
            temperature_end = parameters['temperature_end'] # kelvin
        
        integrator = LangevinMiddleIntegrator(temperature, collision_rate, timestep)
    
        # Set up context
        platform = openmm.Platform.getPlatformByName(platform_name)
        if platform_name in ['CUDA', 'OpenCL']:
            platform.setPropertyDefaultValue('Precision', 'mixed')
        if platform_name in ['CUDA']:
            platform.setPropertyDefaultValue('DeterministicForces', 'true')
        
        context = openmm.Context(system_copy, integrator, platform, {'DisablePmeStream':'true'})
        context.setPeriodicBoxVectors(*system_copy.getDefaultPeriodicBoxVectors())
        context.setPositions(positions)
        context.setVelocitiesToTemperature(temperature)
    
        # Run minimization or MD
        n_steps = parameters['n_steps']
        
        if parameters['EOM'] == 'minimize':
            openmm.LocalEnergyMinimizer.minimize(context, maxIterations=n_steps)
            
        elif parameters['EOM'] == 'MD':
            for _ in tqdm.tqdm_notebook(range(int(n_steps/10000))):
                integrator.step(10000)
              
        elif parameters['EOM'] == 'MD_interpolate':
           
            temperatures = np.linspace(temperature, temperature_end, n_steps)
            n_steps_per_iteration = 1
            for temperature in tqdm.tqdm_notebook(temperatures):
                integrator.setTemperature(temperature)
                integrator.step(n_steps_per_iteration)
    
        else:
            raise Exception("Invalid parameter supplied for 'EOM'")
            
        # Retrieve positions after this stage of equil
        positions = context.getState(getPositions=True).getPositions(asNumpy=True)
        
        # Delete context and integrator 
        del context, integrator, system_copy
        
        elapsed_time = time.time() - initial_time
        print(f"\tStage {i + 1} took {elapsed_time} seconds")
        
    # Save the final equilibrated file
    if filename.endswith('pdb'):
        app.PDBFile.writeFile(topology, positions, open(filename, "w"), keepIds=True)
    elif filename.endswith('cif'):
        app.PDBxFile.writeFile(topology, positions, open(filename, "w"), keepIds=True)

```

with these stages:

```python
stages = [
    {'EOM': 'minimize', 'n_steps': 10000, 'temperature': 300, 'ensemble': None, 'restraint_selection': 'protein and not type H', 'force_constant': 1, 'collision_rate': 2},
    {'EOM': 'MD_interpolate', 'n_steps': 100000, 'temperature': 100, 'temperature_end': 300, 'ensemble': 'NVT', 'restraint_selection': 'protein and not type H', 'force_constant': 1, 'collision_rate': 10},
    {'EOM': 'MD', 'n_steps': 100000, 'temperature': 300, 'ensemble': 'NPT', 'restraint_selection': 'protein and not type H', 'force_constant': 1, 'collision_rate': 10},
    {'EOM': 'MD', 'n_steps': 250000, 'temperature': 300, 'ensemble': 'NPT', 'restraint_selection': 'protein and not type H', 'force_constant': 0.1, 'collision_rate': 2},
    {'EOM': 'minimize', 'n_steps': 10000, 'temperature': 300, 'ensemble': None, 'restraint_selection': 'protein and backbone', 'force_constant': 0.1, 'collision_rate': 2},
    {'EOM': 'MD', 'n_steps': 100000, 'temperature': 300, 'ensemble': 'NPT', 'restraint_selection': 'protein and backbone', 'force_constant': 0.1, 'collision_rate': 2},
    {'EOM': 'MD', 'n_steps': 100000, 'temperature': 300, 'ensemble': 'NPT', 'restraint_selection': 'protein and backbone', 'force_constant': 0.01, 'collision_rate': 2},
    {'EOM': 'MD', 'n_steps': 100000, 'temperature': 300, 'ensemble': 'NPT', 'restraint_selection': 'protein and backbone', 'force_constant': 0.001, 'collision_rate': 2},
    {'EOM': 'MD', 'n_steps': 2500000, 'temperature': 300, 'ensemble': 'NPT', 'restraint_selection': None, 'force_constant': 0, 'collision_rate': 2}
]

```

May be related to https://github.com/openmm/openmm/issues/1728

**Comment by peastman** [comment]
That's basically the equivalent of a segfault.  It means "something went wrong" on the GPU, probably involving memory corruption, but doesn't give much information about exactly what.

A first thing I would try is writing out a serialized State at the end of each stage.  Can you reproduce the error by skipping the first three stages and going straight to stage 4, but starting from the state that was written (in a previous run) at the end of stage 3?

Assuming you can do that, try running the stage 4 simulation inside `cuda-memcheck`.  Does it report any errors?

**Comment by jchodera** [comment]
@zhang-ivy : In your `stages` dict, could you define your quantities as `unit.Quantity` objects so they automatically get converted to OpenMM units via the python API?  I wonder if there's just a unit conversion error that is hiding somewhere that is causing the segfault.

Also, are you using the `force_constant` in this code snippet?

**Comment by jchodera** [comment]
I created a complete example around the code and ran it on `openmmtools.testsystems.SrcExplicit`:
[gentle-equilibration-initial.zip](https://github.com/openmm/openmm/files/7874324/gentle-equilibration-initial.zip)

```
(openmm-7.7) [chodera@lilac:gentle-equilibration]$ python gentle-equilibration.py
Warning: importing 'simtk.openmm' is deprecated.  Import 'openmm' instead.
Executing stage 1
	Stage 1 took 6.7735631465911865 seconds
Executing stage 2
Working... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00
	Stage 2 took 797.7931230068207 seconds
Executing stage 3
Working... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00
	Stage 3 took 834.6498565673828 seconds
Executing stage 4
Working... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--
Traceback (most recent call last):
  File "/lila/home/chodera/gentle-equilibration/gentle-equilibration.py", line 148, in <module>
    run_gentle_equilibration(topology, positions, system, stages, filename, platform_name='CUDA')
  File "/lila/home/chodera/gentle-equilibration/gentle-equilibration.py", line 95, in run_gentle_equilibration
    integrator.step(10000)
  File "/lila/home/chodera/miniconda/envs/openmm-7.7/lib/python3.9/site-packages/openmm/openmm.py", line 11286, in step
    return _openmm.LangevinMiddleIntegrator_step(self, steps)
openmm.OpenMMException: Particle coordinate is nan
```
The issue seems to be that you neglect to retrieve the updated periodic box vectors when retrieving after NPT stages. When you set the positions in the next stage, waters will be erroneously wrapped to collide with others.

Here is an updated version that fixes this, uses units, specifies the timestep as part of the protocol, and cleans up a few things:
[gentle-equilibration-new.zip](https://github.com/openmm/openmm/files/7874428/gentle-equilibration-new.zip)

I'm not sure why the NaN caused by this would trigger a `OpenMMException: Error downloading array buffer: CUDA_ERROR_ILLEGAL_ADDRESS (700)` instead of a correct NaN error. Was this the `openmm-7.7.0` release from conda-forge, the nightly dev build, or a local build from source?

**Comment by dmclark17** [comment]
I haven't confirmed this, but I think incorrect atomic positions—specifically, a large number of overlapping atoms—can cause invalid memory accesses in the `findBlocksWithInteractions` kernel. 

https://github.com/openmm/openmm/blob/f2cdf70afc3ec4f4f0a4fc3aa5ca4761fc994cf1/platforms/cuda/src/kernels/findInteractingBlocks.cu#L347-L352

If the number of interactions is between 2^31 and 2^32-1, when the interaction count is cast to a signed integer, I believe it will produce a negative result. This will pass the maxTiles check and lead to an invalid read.

**Comment by peastman** [comment]
Good point!  It looks like we probably want to change `tileStartIndex`, `newTileStartIndex`, and `tilesToStore` to all be unsigned?

**Comment by zhang-ivy** [comment]
>Also, are you using the force_constant in this code snippet?

Oops, I accidentally forgot to include that code. 

>The issue seems to be that you neglect to retrieve the updated periodic box vectors when retrieving after NPT stages. When you set the positions in the next stage, waters will be erroneously wrapped to collide with others.

Ah thanks! Your updated code (plus my restraint force code) now works fine on RBD:ACE2!

>Was this the openmm-7.7.0 release from conda-forge, the nightly dev build, or a local build from source?
Its from a nightly build.

**Comment by dmclark17** [comment]
> Good point! It looks like we probably want to change tileStartIndex, newTileStartIndex, and tilesToStore to all be unsigned?

I think so! And then in the nonbonded kernel, we may need to convert `pos` and `end` to unsigned integers as well:

https://github.com/openmm/openmm/blob/f2cdf70afc3ec4f4f0a4fc3aa5ca4761fc994cf1/platforms/cuda/src/kernels/nonbonded.cu#L386


Although with that many interactions, I imagine most GPUs would run out of memory.

**Comment by jchodera** [comment]
Would it make sense to do some kind of sanity check instead?

**Comment by peastman** [comment]
The nans are the cause of the out of memory.  Under normal circumstances, there's plenty of memory.  But when all the coordinates become nan, it concludes that every atom interacts with every other, so it builds a huge neighbor list.

I think this is only happening because we compile with `--use_fast_math`.  The checks are written to handle nans correctly:

https://github.com/openmm/openmm/blob/f2cdf70afc3ec4f4f0a4fc3aa5ca4761fc994cf1/platforms/cuda/src/kernels/findInteractingBlocks.cu#L321-L323

If any coordinate is nan then `halfDist2` will be nan, so the check should fail and `interacts` will not get set.  But we told the compiler it can assume nothing will ever be nan, so it doesn't implement the check correctly.  At least, I assume that's what's happening.

I think there are two steps to the fix.  First, make sure we won't ever get invalid memory access in either kernel, even when things are nan.  And second, whenever we see the neighbor list is too small (in `CudaNonbondedUtilities::updateNeighborListSize()`), we should probably check for any nan coordinates before trying to allocate more memory for it.

**Comment by dmclark17** [comment]
I experimented with changing `tileStartIndex`, `newTileStartIndex`, and `tilesToStore` to unsigned integers; however, it didn't completely resolve the invalid memory access issue. 

I believe we can get an invalid memory access if we overflow the unsigned integer as well. If `newTileStartIndex+tilesToStore` overflows, then this check will pass:

https://github.com/openmm/openmm/blob/f2cdf70afc3ec4f4f0a4fc3aa5ca4761fc994cf1/platforms/cuda/src/kernels/findInteractingBlocks.cu#L350

But, thread 0 will attempt to write at `newTileStartIndex` which has not overflown:

https://github.com/openmm/openmm/blob/f2cdf70afc3ec4f4f0a4fc3aa5ca4761fc994cf1/platforms/cuda/src/kernels/findInteractingBlocks.cu#L352

There may be other cases that can cause an invalid read—I will look into ways to prevent these. Should we keep track of if `interactionCount[0]` overflows?

**Comment by peastman** [comment]
I was expecting that would happen.  The reason is that even after it passes `maxTiles`, it keeps running and keeps incrementing `interactionCount`.  I think the solution is to make it stop immediately:

```c++
if (newTileStartIndex+tilesToStore > maxTiles)
    return
```

**Comment by dmclark17** [comment]
That makes sense to me. Another potential issue is that if the number of particles is greater than 2^21 the `totalTiles` variable will overflow here:

https://github.com/openmm/openmm/blob/3516752a16afca2d74a166d5b27456da08f74c20/platforms/cuda/src/CudaNonbondedUtilities.cpp#L434

I think `totalTiles` should probably become a long since it only takes 2^21.5 particles to overflow an unsigned integer.

**Comment by peastman** [comment]
Although there's a problem with what I suggested above.  In the normal case where the neighbor list is just a little too small, we do want to keep going so we can get an accurate count of how big it needs to be.  Maybe just apply a sanity check and stop when the number of tiles exceeds 2^31?

Let me see what I can get working.

**Comment by peastman** [comment]
After investigating some more, I think we're barking up the wrong tree.  The problem above was reported in a system with 200,000 atoms, meaning about 6000 atom blocks.  Even if every atom were within the cutoff distance of every other atom, that would still only be about 18 million tiles.  To reach the limit of a signed int, you would need about 2 million atoms.

So the problem must be something different than what we've been discussing.

**Comment by peastman** [comment]
Here's the problem.  To make the neighbor list bigger, it resizes the `interactingTiles` and `interactingAtoms` arrays.

https://github.com/openmm/openmm/blob/446aaeb47db6dc2d2fe156bc4484847d6281e209/platforms/cuda/src/CudaNonbondedUtilities.cpp#L437-L438

The latter gets resized to 32 times the number of tiles.  The argument to `resize()` is an int.  Long before the number of tiles reaches the limit of a signed int, the size of that array overflows it.  It gets resized to the wrong value, which leads to an illegal access on the next call to `findInteractingBlocks()`.

I think the solution is to change the size of arrays from int to size_t.

**Comment by peastman** [comment]
The fix is in #3428.
